src/Base/Tools.cpp

```diff
@@ -33,130 +33,217 @@
-        str << base_name;
-        if (padding > 0) {
-            str.fill('0');
-            str.width(padding);
-        }
-        str << Base::string_comp::increment(num_suffix);
-        return str.str();
+    else {
+        // value is in the moddle of the span, so we must split it.
+        etype firstReplacement(at->first, value - at->first);
+        etype secondReplacement(value + 1, at->second - ((value + 1) - at->first));
+        // Because erase returns the iterator after the erased element, and insert returns the
+        // iterator for the inserted item, we want to insert secondReplacement first.
+        Spans.insert(Spans.insert(Spans.erase(at), secondReplacement), firstReplacement);
     }
+}
+bool Base::UniqueNameManager::PiecewiseSparseIntegerSet::Contains(uint value) const
+{
+    iterator at = Spans.lower_bound(etype(value, 1));
+    return at != Spans.end() && at->first <= value;
+}
 
-private:
-    std::string num_suffix;
-    std::string base_name;
-    int padding;
-};
-
-}  // namespace Base
-
-std::string
-Base::Tools::getUniqueName(const std::string& name, const std::vector<std::string>& names, int pad)
+std::tuple<uint, uint> Base::UniqueNameManager::decomposeName(const std::string& name,
+                                                              std::string& baseNameOut,
+                                                              std::string& nameSuffixOut) const
 {
-    if (names.empty()) {
-        return name;
+    std::string::const_iterator suffixStart = GetNameSuffixStartPosition(name);
+    nameSuffixOut = name.substr(suffixStart - name.cbegin());
+    std::string::const_iterator digitsStart = suffixStart;
+    while (digitsStart > name.cbegin() && isdigit(digitsStart[-1])) {
+        --digitsStart;
     }
-
-    Base::unique_name unique(name, names, pad);
-    return unique.get();
+    baseNameOut = name.substr(0, digitsStart - name.cbegin());
+    uint digitCount = suffixStart - digitsStart;
+    if (digitCount == 0) {
+        // No digits in name
+        return std::tuple<uint, uint> {0, 0};
+    }
+    else {
+        return std::tuple<uint, uint> {
+            digitCount,
+            std::stol(name.substr(digitsStart - name.cbegin(), digitCount))};
+    }
+}
+void Base::UniqueNameManager::addExactName(const std::string& name)
+{
+    std::string baseName;
+    std::string nameSuffix;
+    uint digitCount;
+    uint digitsValue;
+    std::tie(digitCount, digitsValue) = decomposeName(name, baseName, nameSuffix);
+    baseName += nameSuffix;
+    auto baseNameEntry = UniqueSeeds.find(baseName);
+    if (baseNameEntry == UniqueSeeds.end()) {
+        // First use of baseName
+        baseNameEntry = UniqueSeeds
+                            .insert(std::pair<std::string, std::vector<PiecewiseSparseIntegerSet>>(
+                                baseName,
+                                std::vector<PiecewiseSparseIntegerSet>()))
+                            .first;
```

Comment:
The idea was to make it easier to read.
```c++
    if (baseNameEntry == UniqueSeeds.end()) {
        // First use of baseName
        std::pair<std::string, std::vector<PiecewiseSparseIntegerSet>> seed = {
            baseName, {}};
        baseNameEntry = UniqueSeeds.insert(seed).first;
    }
```

Or even better with:
```c++
using Seed = std::pair<std::string, std::vector<PiecewiseSparseIntegerSet>>; 
```

Then it could be written like this:
```c++
    if (baseNameEntry == UniqueSeeds.end()) {
        // First use of baseName
        auto seed = Seed{baseName, {}};
        baseNameEntry = UniqueSeeds.insert(seed).first;
    }
```

But you are right that neither of these are good as it will create the pair, even if it isn't needed just like the suggestion you did. By changing insert to [emplace](https://en.cppreference.com/w/cpp/container/map/emplace) fixes that:

```c++
    if (baseNameEntry == UniqueSeeds.end()) {
        // First use of baseName
        baseNameEntry = UniqueSeeds.emplace(baseName, {}).first;
    }
```

Reply:
Hmm, my compiler did not seem to like that last one, I had to explicitly name the vector type.
Still a lot easier to read though.